File: NSTViewWarmuper/ViewController.m
//
//  ViewController.m
//  NSTViewWarmuper
//
//  Created by Timur Bernikowich on 2/6/17.
//  Copyright © 2017 Timur Bernikovich. All rights reserved.
//

#import "ViewController.h"
#import "WebViewController.h"

typedef UIViewController * (^ControllerConfigurationCreationBlock)();

@interface ControllerConfiguration : NSObject

@property (nonatomic) NSString *title;
@property (nonatomic) ControllerConfigurationCreationBlock creationBlock;

+ (instancetype)objectWithTitle:(NSString *)title creationBlock:(ControllerConfigurationCreationBlock)block;

@end

@implementation ControllerConfiguration

+ (instancetype)objectWithTitle:(NSString *)title creationBlock:(ControllerConfigurationCreationBlock)block
{
    ControllerConfiguration *object = [self new];
    object.title = title;
    object.creationBlock = block;
    return object;
}

@end

@interface ViewController () <UITableViewDataSource, UITableViewDelegate>

@property (nonatomic) UITableView *tableView;
@property (nonatomic) NSArray<ControllerConfiguration *> *configurationArray;

@end

@implementation ViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    self.tableView = [[UITableView alloc] initWithFrame:self.view.bounds style:UITableViewStylePlain];
    self.tableView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    [self.tableView registerClass:UITableViewCell.class forCellReuseIdentifier:NSStringFromClass(UITableViewCell.class)];
    [self.view addSubview:self.tableView];
    
    NSMutableArray<ControllerConfiguration *> *array = [NSMutableArray new];
    [array addObject:[ControllerConfiguration objectWithTitle:@"UIWebView" creationBlock:^UIViewController *{
        WebViewController *controller = [WebViewController new];
        controller.webKit = NO;
        controller.warmup = NO;
        return controller;
    }]];
    [array addObject:[ControllerConfiguration objectWithTitle:@"UIWebView + Warmuper" creationBlock:^UIViewController *{
        WebViewController *controller = [WebViewController new];
        controller.webKit = NO;
        controller.warmup = YES;
        return controller;
    }]];
    [array addObject:[ControllerConfiguration objectWithTitle:@"WKWebView" creationBlock:^UIViewController *{
        WebViewController *controller = [WebViewController new];
        controller.webKit = YES;
        controller.warmup = NO;
        return controller;
    }]];
    [array addObject:[ControllerConfiguration objectWithTitle:@"WKWebView + Warmuper" creationBlock:^UIViewController *{
        WebViewController *controller = [WebViewController new];
        controller.webKit = YES;
        controller.warmup = YES;
        return controller;
    }]];
    self.configurationArray = array;
}

#pragma mark - Table View

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.configurationArray.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:NSStringFromClass(UITableViewCell.class)];
    cell.textLabel.text = self.configurationArray[indexPath.row].title;
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    [self.navigationController pushViewController:self.configurationArray[indexPath.row].creationBlock() animated:YES];
}

@end

File: Example/WebViewWarmUper/ViewController.swift
//
//  Created by Timur Bernikovich on 07/03/2019.
//  Copyright © 2019 Timur Bernikovich. All rights reserved.
//

import UIKit

struct WebViewControllerConfiguration {
    let title: String
    let warmUp: Bool

    func createWebViewController() -> UIViewController {
        let controller = WebViewController(warmUp: warmUp)
        controller.title = title
        return controller
    }
}

class ViewController: UIViewController {

    let tableView = UITableView(frame: .zero, style: .plain)
    let configurationCollection: [WebViewControllerConfiguration] = {
        var collecton: [WebViewControllerConfiguration] = []
        
        collecton.append(WebViewControllerConfiguration(title: "WKWebView", warmUp: false))
        collecton.append(WebViewControllerConfiguration(title: "WKWebView + WarmUper", warmUp: true))

        return collecton
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        tableView.frame = view.bounds
        tableView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        tableView.delegate = self
        tableView.dataSource = self
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "Cell")
        view.addSubview(tableView)
    }
}

extension ViewController: UITableViewDelegate, UITableViewDataSource {
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return configurationCollection.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
        cell.textLabel?.text = configurationCollection[indexPath.row].title
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        navigationController?.pushViewController(configurationCollection[indexPath.row].createWebViewController(), animated: true)
    }
}
